nixos/modules/services/web-apps/mastodon.nix

```diff
@@ -0,0 +1,475 @@
+      enable = lib.mkEnableOption "Mastodon federated social network";
+
+      configureNginx = lib.mkOption {
+        description = ''
+          Configure nginx as reverse proxy for mastodon
+          Note that this makes some assumptions on your setup, and sets settings that will
+          affect other virtualHosts running on your nginx instance, if any. 
+          Alternatively you can configure a reverse-proxy of your choice to serve these paths:
+
+          / -> $(nix-instantiate --eval '&lt;nixpkgs&gt;' -A mastodon.outPath)/public
+          / -> 127.0.0.1:{{ webPort }} (if there was no file in the directory above)
+          /system/ -> /var/lib/mastodon/public-system/
+          /api/v1/streaming/ -> 127.0.0.1:{{ streamingPort }}
+
+          Make sure that websockets are forwarded properly. You might want to set up caching
+          of some requests, take a look at mastodon's provided nginx configuration at
+          https://github.com/tootsuite/mastodon/blob/master/dist/nginx.conf
+        '';
+        type = lib.types.bool;
+        default = false;
+      };
+
+      user = lib.mkOption {
+        description = ''
+          User under which mastodon runs
+          If it is set to "mastodon", a user will be created.
+        '';
+        type = lib.types.str;
+        default = "mastodon";
+      };
+
+      group = lib.mkOption {
+        description = ''
+          Group under which mastodon runs
+          If it is set to "mastodon", a group will be created.
+        '';
+        type = lib.types.str;
+        default = "mastodon";
+      };
+
+      streamingPort = lib.mkOption {
+        description = "TCP port used by the mastodon-streaming service";
+        type = lib.types.port;
+        default = 55000;
+      };
+
+      webPort = lib.mkOption {
+        description = "TCP port used by the mastodon-web service";
+        type = lib.types.port;
+        default = 55001;
+      };
+
+      sidekiqPort = lib.mkOption {
+        description = "TCP port used by the mastodon-sidekiq service";
+        type = lib.types.port;
+        default = 55002;
+      };
+
+      vapidPublicKeyFile = lib.mkOption {
+        description = ''
+          Path to file containing the public key used for Web Push Voluntary Application Server Identification
+          A keypair can be generated by running
+          cd ${pkgs.mastodon}; bin/rake webpush:generate_keys
+        '';
+        default = "/var/lib/mastodon/secrets/vapid-public-key";
+        type = lib.types.str;
+      };
+
+      localDomain = lib.mkOption {
+        description = "The domain serving your Mastodon instance";
+        default = "social.example.org";
+        type = lib.types.str;
+      };
+
+      secretKeyBaseFile = lib.mkOption {
+        description = ''
+          Path to file containing the secret key base
+
+          Can be generated by running
+          cd $(nix-instantiate --eval '&lt;nixpkgs&gt;' -A mastodon.outPath); bin/rake secret
```

Comment:
`nix-instantiate --eval '<nixpkgs>' -A mastodon.outPath` just tells you what the path of the Mastodon package is. It doesn't actually build the package, so if the package isn't already in the Nix store, as might be the case if you're just getting started setting up this module, this command won't work.

The best replacement I've come up with is: `nix build -f '&lt;nixpkgs&gt;' mastodon; cd result; bin/rake secret`

See my branch [60788-tootctl](https://github.com/happy-river/nixpkgs/tree/60788-tootctl) for a commit where I've made this change in the several places where it is needed, added mention that the keys will be generated automatically if they are not supplied, and also edited most of the other option descriptions for formatting and clarity.